python/src/iceberg/types.py

```diff
@@ -190,13 +242,25 @@ def __init__(self, element: NestedField):
     def element(self) -> NestedField:
         return self._element_field
 
-    def __eq__(self, other):
-        if type(self) is type(other):
-            return self.element == other.element
-        return False
 
+class MapType(IcebergType):
+    """A map type in Iceberg
+
+    Example:
+        >>> MapType(
+                NestedField(True, 1, "required_field", StringType()),
```

Comment:
Map keys are not allowed to be null. If we are creating a map by passing two `NestedField` instances, then we should check that. Same thing with names. Map key fields must be named "key" and map value fields must be named "value".

This is why the Java API passes the types and field IDs instead of accepting nested fields:

```java
MapType stringToIntMap = MapType.ofRequired(1, 2, StringType.get(), IntegerType.get());
// Also defines `ofOptional`
```